### ISSUE TITLE
False positive when Lock invoked via method

### ISSUE THREAD

**Issue by jacob-pro** [spotbugs/spotbugs#3516]
It seems that when a Java concurrent lock is used indirectly, such as this example with an AutoClosable wrapper, then it causes false positives AT_NONATOMIC_OPERATIONS_ON_SHARED_VARIABLE and AT_STALE_THREAD_WRITE_OF_PRIMITIVE.

```java
public class Demo {

    private final CloseableReentrantLock lock = new CloseableReentrantLock();

    private final ScheduledExecutorService executor = Executors.newScheduledThreadPool(5);

    private int consecutiveUnsuccessfulCalls;

    /**
     * Start the scheduled task.
     */
    public void start() {
        executor.schedule(this::doTask, 5, MILLISECONDS);
    }

    void doTask()  {
        try (var ignored = lock.open()) {
            try {
                Thread.sleep(1);
                consecutiveUnsuccessfulCalls = 0;
            } catch (Exception e) {
                consecutiveUnsuccessfulCalls++;
            } finally {
                scheduleNextCall();
            }
        }
    }

    private void scheduleNextCall() {
        if (consecutiveUnsuccessfulCalls < 5) {
            executor.schedule(this::doTask, 5, MILLISECONDS);
        }
    }

    static class CloseableReentrantLock extends ReentrantLock {

        Guard open() {
            this.lock();
            return new Guard();
        }

        class Guard implements AutoCloseable {

            @Override
            public void close() {
                CloseableReentrantLock.this.unlock();
            }
        }
    }
}
```

```
[ERROR] Medium: Operation on the "consecutiveUnsuccessfulCalls" shared variable in "Demo" class is not atomic [org.example.Demo] At Demo.java:[line 30] AT_NONATOMIC_OPERATIONS_ON_SHARED_VARIABLE
[ERROR] Medium: Shared primitive variable "consecutiveUnsuccessfulCalls" in one thread may not yield the value of the most recent write from another thread [org.example.Demo] At Demo.java:[line 28] AT_STALE_THREAD_WRITE_OF_PRIMITIVE
```

If the method is written to invoke the lock directly, then there is no error reported:

```java
    void doTask()  {
        lock.lock();
        try {
            try {
                Thread.sleep(1);
                consecutiveUnsuccessfulCalls = 0;
            } catch (Exception e) {
                consecutiveUnsuccessfulCalls++;
            } finally {
                scheduleNextCall();
            }
        } finally {
            lock.unlock();
        }
    }
```

**Comment by gtoison** [comment]
Hum, that's probably because SpotBugs doesn't know `CloseableReentrantLock`, is that a class you have written?

**Comment by jacob-pro** [comment]
yes, you can see the class included in the example code above. The problem is the same any time `.lock()` is not called directly, but via another method

**Comment by gtoison** [comment]
From what I can see SpotBug's LockAnalysis looks for calls to zero-arg methods named "lock" (or "lockInterruptibly") and "unlock", and they must be on the same object. It does not attempt to analyse what the methods are actually doing. 

In your case the lock calls are named open/close and are not on the same object (the lock vs. the guard), indeed Spotbugs is not smart enough to see that.

Fixing that false positive does not seem trivial unfortunately
